### ISSUE TITLE
Password encryption for adminpassword and others

### ISSUE THREAD

**Issue by zephir89** [simplesamlphp/simplesamlphp#405]
Hi,
We want that the passwords stored in config files, like adminpassword or other passwords in other modules (ldap, sql, etc), must be encrypted using asymmetric encryption.

**Comment by pmeulen** [comment]
Hi zephir89,
What problem do you want to solve by encrypting the passwords stored in the config files?
Where and how do you want to store the decryption key?

**Comment by zephir89** [comment]
Hi pmeulen, 
For security reason, I don't want to store cleartext password in config files.

**Comment by jaimeperez** [comment]
Hi @zephir89!

As @pmeulen was asking, where are you going to store the decryption key for the passwords in your config file? In the very same config file? Or are you going to use an HSM (which is currently not supported by SimpleSAMLphp?

Storing your passwords encrypted (it doesn't really matter whether you use symmetric or asymmetric encryption) is a very bad idea in general. If you are worried about the admin password, you can hash it with the `bin/pwgen.php` script. LDAP passwords can (and should) be hashed too, and that's transparent for SimpleSAMLphp. Read more [here](https://simplesamlphp.org/docs/stable/authcrypt:authcrypt).

Now, if you are worried about storing passwords in a database, you will need to use per-user, per-password salts, and store the hash of the salt and the password itself. Read more about properly storing passwords [here](https://crackstation.net/hashing-security.htm). If you are really interested in this topic, I recommend you to read [this article](http://arstechnica.com/security/2016/06/how-linkedins-password-sloppiness-hurts-us-all/) for an insight on what are the consequences of bad practices, and how to manage accounts and passwords properly in general.

**Comment by zephir89** [comment]
Hello @jaimeperez. Thanks for your reply about hashing admin password.
For LDAP instead, we have read php documentation and it's not possible to use hashed password in ldap_bind function. (http://stackoverflow.com/questions/20821571/php-ldap-bind-using-a-hash)
For this reason my idea was to encrypt LDAP password in config file (i.e. using the same key used for encrypt SAMLRequest) and then decrypt it on LDAP binding.

Thank you.

**Comment by jaimeperez** [comment]
Hi!

It's not that it's not possible to use hashed passwords when binding to an LDAP server, it's just not necessary. LDAP itself stores the passwords hashed, not in plaintext. If you are concerned about the password traveling through the (probably internal) network in plaintext, what you need to do is use **LDAPS** (TCP port 636) instead of plain LDAP.

I'm not sure however that I understand what you mean. You don't need to have any *LDAP password* in your config file, as you will perform the binding with the username and password provided by the user himself.

Now, if you want to encrypt the passwords in the LDAP, you cannot use the standard password field and you will need something else. This also implies that you cannot use the LDAP for authentication (since you can no longer *bind* to it), so you will need to implement your own LDAP authentication that does **not** do an `ldap_bind`. In that case, you will still have to bind to the LDAP server, since that's how the LDAP protocol works. Either you bind anonymously (in which case the encrypted passwords would be available for everyone who could reach the directory), or with a user with administrative privileges that can read all user entries (and whose password you will send in plaintext over the network when binding, anyway, so again you need to use LDAPS, which would solve the initial problem since the beginning).

Additionally, you would need to come up with your own crypto (unless you plan to use some standard like JWE, which SimpleSAMLphp does not know how to handle), which is always a very bad idea.

So to sum up, if your concern is the security of your user accounts, using LDAP with hashed (standard) passwords over a secured channel (LDAPS) is perfectly fine. That's what everybody using LDAP is doing, and there is no security issues in there. Introducing your own authentication mechanism together with your own way to store the passwords is actually compromising the security of your deployment and making it more insecure than using LDAP authentication out of the box.

**Comment by zephir89** [comment]
Thank you @jaimeperez! We have understood what you said. Maybe we are explaining bad... We'll try to do better. In ldap.php config template file there is:
```
/* The username & password the simpleSAMLphp should bind as before searching. If this is left
* as NULL, no bind will be performed before searching.
*/
'auth.ldap.search.username' => NULL,
'auth.ldap.search.password' => NULL,
```
Like for the admin password in the config.php file we would like to avoid the use of a plaintext password in this file. So, it isn't a problem concerned with the way LDAP stores user password or performing the binding with user credentials. Is it possible?
Thank you!

**Comment by jaimeperez** [comment]
Hi again!

Now I see where the confusion is coming from. First of all, ignore the `config-templates/ldap.php` file. It's a leftover from the past and I'm going to remove it right now. The LDAP authentication is not configured like that (you need to configure an authentication source instead) and the configuration options don't have these names either.

Regarding the confusion itself, what you are reading is a particular feature that you **can** use with LDAP. The **can** in the last sentence is the key, as you don't **need** to do that. That will only be necessary if your users are scattered all over the directory and you don't have a base `dn` where you have them.

If that's really your case and you **need** to search for your users, you just need to protect your configuration file the same way you would need to protect it already for other secrets, so it's not a big deal. A good example is the *salt*, which is a key secret that should never be disclosed.

In any case, let's imagine you encrypt this particular password with the private key you are using to sign SAML messages. What's the difference then between the private key and the `config.php` or `authsources.php` files? What's the benefit of encrypting a secret with another secret that is available next to the secret you are trying to protect?

The only way this could make sense is if you were using an HSM to store your private key, and make SimpleSAMLphp use it to sign / decrypt whenever you need it. But SimpleSAMLphp doesn't support HSMs right now anyway, so...
